titus-server-master/src/main/java/com/netflix/titus/master/kubernetes/pod/affinity/DefaultPodAffinityFactory.java

```diff
@@ -285,13 +293,52 @@ private void processZoneConstraints() {
             }
         }
 
+        private void processRelocationAffinity() {
+            if (!featureConfiguration.isRelocationBinpackingEnabled() || !JobManagerUtil.getRelocationBinpackMode(job).isPresent()) {
+                return;
+            }
+            getPodAffinity().addPreferredDuringSchedulingIgnoredDuringExecutionItem(
+                    new V1WeightedPodAffinityTerm()
+                            .weight(RELOCATION_AFFINITY_WEIGHT)
+                            .podAffinityTerm(new V1PodAffinityTerm()
+                                    .labelSelector(new V1LabelSelector()
+                                            .addMatchExpressionsItem(new V1LabelSelectorRequirement()
+                                                    .key(KubeConstants.POD_LABEL_RELOCATION_BINPACK)
+                                                    .operator(KubeConstants.SELECTOR_OPERATOR_EXISTS)
+                                            )
+                                    )
+                                    .topologyKey(KubeConstants.NODE_LABEL_MACHINE_ID)
+                            )
```

Comment:
This does not distinguish between different jobs, does it?
We must ensure that job spreading is disabled. This is controlled independently in `DefaultTopologyFactory`. Maybe we should change that implementation as well and check `isRelocationBinpackingEnabled` there.
Or the goal here is to limit number of machines used, but preserve the original spreading behavior?

Reply:
I wanted to preserve the original spreading behavior, but favor binpacking when possible (when multiple choices given the current spreading behavior is maintained). Let me at least review the weights in both cases